### PR TITLE
tls support: Added support for certless communcation

### PR DESCRIPTION
**Pull request by alorbach**
Added support for Anon ciphers (certless communcation) in gtls and ossl driver. 
Certless ciphers are only being added when authmode is set to ANON. 
- Server and Client code can now run without CA, cert and key. 
- Also added new tests for certless server, client and both - for gtls and ossl.

### CODE REVIEW

**Comment by rgerhards** [comment]
general question before I go into a real review: shouldn't we require an explicit opt-in into certless mode? As we know, this is pretty insecure, so I would tend to let the code still err out if not explicitly selected. Folks strongly tend to ignore warning messages.

The opt-in could be via `authmode="certless"` - just as an example.

@davidelang @Whissi @mbiebl @jvymazal what do you think?

File: runtime/nsd_ossl.c [hunk 1]
```diff
@@ -85,6 +85,11 @@ static int bGlblSrvrInitDone = 0;	/**< 0 - server global init not yet done, 1 -
 /* Main OpenSSL CTX pointer */
 static SSL_CTX *ctx;
 
+/* Static Helper variables for CERT status */
+short bHaveCA;
+short bHaveCert;
+short bHaveKey;
```
Comment (rgerhards) [inline runtime/nsd_ossl.c]: the comment says these are meant as `static`...

File: tests/sndrcv_tls_certless.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+# alorbach, 2019-01-16
+# testing sending and receiving via TLS with anon auth using bare ipv4, no SNI
```
Comment (rgerhards) [inline tests/sndrcv_tls_certless.sh]: I guess this comment is incorrect... - or what does "bare ipv4, no SNI" mean in this context?

File: tests/sndrcv_tls_certless_serveronly.sh [hunk 3]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+# alorbach, 2019-01-16
+# testing sending and receiving via TLS with anon auth using bare ipv4, no SNI
```
Comment (rgerhards) [inline tests/sndrcv_tls_certless_serveronly.sh]: I guess this comment is incorrect... - or what does "bare ipv4, no SNI" mean in this context?

File: tests/sndrcv_tls_ossl_certless.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
+#!/bin/bash
+# alorbach, 2019-01-16
+# testing sending and receiving via OpenSSL TLS with anon auth using bare ipv4, no SNI
```
Comment (rgerhards) [inline tests/sndrcv_tls_ossl_certless.sh]: I guess this comment is incorrect... - or what does "bare ipv4, no SNI" mean in this context?

File: tests/sndrcv_tls_ossl_certless_clientonly.sh [hunk 5]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+# alorbach, 2019-01-16
+# testing sending and receiving via OpenSSL TLS with anon auth using bare ipv4, no SNI
```
Comment (rgerhards) [inline tests/sndrcv_tls_ossl_certless_clientonly.sh]: I guess this comment is incorrect... - or what does "bare ipv4, no SNI" mean in this context?

File: tests/sndrcv_tls_ossl_certless_clientonly.sh [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+# This file is part of the rsyslog project, released  under ASL 2.0
+
+# uncomment for debugging support:
+. ${srcdir:=.}/diag.sh init
+export NUMMESSAGES=25000
+# start up the instances
+#export RSYSLOG_DEBUG="debug nostdout noprintmutexaction"
```
Comment (rgerhards) [inline tests/sndrcv_tls_ossl_certless_clientonly.sh]: in general, the comment lines here and in other tests would better be in proper order (albeit we can guess what is meant).

File: tests/sndrcv_tls_ossl_certless_serveronly.sh [hunk 7]
```diff
@@ -0,0 +1,71 @@
+#!/bin/bash
+# alorbach, 2019-01-16
+# testing sending and receiving via OpenSSL TLS with anon auth using bare ipv4, no SNI
```
Comment (rgerhards) [inline tests/sndrcv_tls_ossl_certless_serveronly.sh]: I guess this comment is incorrect... - or what does "bare ipv4, no SNI" mean in this context?

File: runtime/nsd_ossl.c (hunk 1 above)
Comment (alorbach) [reply runtime/nsd_ossl.c]: correct they should be static

File: tests/sndrcv_tls_certless.sh (hunk 2 above)
Comment (alorbach) [reply tests/sndrcv_tls_certless.sh]: arg copy & paste ;)

File: tests/sndrcv_tls_ossl_certless_clientonly.sh (hunk 6 above)
Comment (alorbach) [reply tests/sndrcv_tls_ossl_certless_clientonly.sh]: Copy & paset from original test

**Comment by rgerhards** [comment]
> The draft actually doesn't menation ANON at all:
> https://tools.ietf.org/html/draft-ietf-syslog-transport-tls-12

Actually, the doc should change to say RFC5425. And even more so, explicit anon was dropped from the RFC for security reasons. Let me dig out a somewhat older issue where I describe this all at length.

But in "anon" mode I think we still validate the certs - don't we? In "certless" we obviously cannot do that, so it is definitely the least secure mode of all.

**Comment by rgerhards** [comment]
@alorbach you need to run the ossl tests only when ossl is enabled

**Comment by rgerhards** [comment]
This pull request **introduces 2 alerts** when merging cd37cfde1a88a1712edd25160496c3249d545541 into 734ea72c448d1baf438d5603c95cca1bc726e1d6 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-a22ea1d967e544246364f5e04002570036147f3c)

**new alerts:**

* 2 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 2 alerts** when merging dd792d164b9f5e887c2d394beb2623fc626dd016 into 734ea72c448d1baf438d5603c95cca1bc726e1d6 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-32139da46ea1b6369ac27e20a7711429a149370a)

**new alerts:**

* 2 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Whissi** [comment]
Just to get it right: The RFC I am aware of is about allowing TLS without the need of maintaining an own CA and create client certificates. However, drafts say anon cipher will still require the "server" to maintain and authorize clients. How is this addressed in the 'certless' mode your are currently implementing?

**Comment by rgerhards** [comment]
This pull request **introduces 3 alerts** when merging f8628fba6342310b088b13a9d3317026c42dd3e6 into 734ea72c448d1baf438d5603c95cca1bc726e1d6 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-a6869e23af1f2666b3448c58f5c197d0ab37796f)

**new alerts:**

* 3 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 3 alerts** when merging a66395a04fc157f2d116c29e3c92fdfaa6be343e into 734ea72c448d1baf438d5603c95cca1bc726e1d6 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-819bb4c5b6300b529e242a558797180b43d00b56)

**new alerts:**

* 3 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 3 alerts** when merging 1fb5bbf0f614f32425f2b9915d49c4404123e8d8 into 734ea72c448d1baf438d5603c95cca1bc726e1d6 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-622bccf05bcdc4f810d1c22dda45cc768be94584)

**new alerts:**

* 3 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 4 alerts** when merging 89b90975c1dfada87c1036c3e4dd356341752729 into 734ea72c448d1baf438d5603c95cca1bc726e1d6 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-f7c23faf3f91b00d94e3ea0304c3fe2ff466f795)

**new alerts:**

* 4 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 4 alerts** when merging 8860454ffca89215446437e9adf8f178660b157e into c533db69e447ceb1e56ca924d251cf36a0c05bd7 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-4fe24b25705658415328921facb77ee3b1e07a8a)

**new alerts:**

* 4 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 4 alerts** when merging c86e39f987a20ad6bcae5e3d4606c25971ed4a9f into c533db69e447ceb1e56ca924d251cf36a0c05bd7 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-e73fc5b5ae85bf7d9a3c5d431c2666f91f4092e1)

**new alerts:**

* 4 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 4 alerts** when merging d9802b7fccee8062607a906f1907bcd6af217d11 into c533db69e447ceb1e56ca924d251cf36a0c05bd7 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-eb67bf72fa7907d1b5549af26b2f12b041a653ed)

**new alerts:**

* 4 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 4 alerts** when merging 9ab7be955bf48db8c61daeddbf744ee63b5f2706 into c533db69e447ceb1e56ca924d251cf36a0c05bd7 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-6c539068372475c01e3df8a088a0c3e2d2d5ded9)

**new alerts:**

* 4 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 58a37cd12e9b82e6341b18f6cfad96c230ac66b3 into c533db69e447ceb1e56ca924d251cf36a0c05bd7 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-254f67c9563a5aed886f65edd9683b0af51ec653)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 346f276dd92c463364d148684ca34f0b54c297fd into 017815e2cfeaf9f02b68c042ab2e52202d610c4a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-dde2fc58fb298e1047951b8d5370da0a68d21371)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 3e536974ad1a8588d9dd51fb811fb5bcf80c69ce into 017815e2cfeaf9f02b68c042ab2e52202d610c4a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-1751da043fe2d278ce412a95517a7c3b0a63247b)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging ba48755a1d48be1d0a15fdecb92393c636ad49d5 into 017815e2cfeaf9f02b68c042ab2e52202d610c4a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-7d90595ebb9bd6615d4a05e3d5ba75774b86dfb1)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 57bbbb5b4f905d2c03b05499c4b6ecdf498158b3 into 16c3cfab69ccc11c200d4e02d9d83b8efb0f9077 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-a436a1c634ce138cb13609f4d026de89efadde66)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by rgerhards** [comment]
This pull request **introduces 1 alert** when merging 30cf761692f347bc40c387da1b5a49ed5bb68202 into 382dc8458af3b68d85a2733973b932e7957cf782 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-974dbb47eb5a523987bbd53893d659f41f13b84e)

**new alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: tests/imtcp-tls-basic.sh [hunk 8]
```diff
@@ -2,7 +2,7 @@
 # added 2011-02-28 by Rgerhards
 # This file is part of the rsyslog project, released under ASL 2.0
 . ${srcdir:=.}/diag.sh init
-export NUMMESSAGES=50000
+export NUMMESSAGES=1000
```
Comment (rgerhards) [inline tests/imtcp-tls-basic.sh]: we need to get the old value here

File: tests/imtcp-tls-ossl-basic.sh [hunk 9]
```diff
@@ -2,7 +2,7 @@
 # added 2018-04-27 by alorbach
 # This file is part of the rsyslog project, released under ASL 2.0
 . ${srcdir:=.}/diag.sh init
-export NUMMESSAGES=10000
+export NUMMESSAGES=1000
```
Comment (rgerhards) [inline tests/imtcp-tls-ossl-basic.sh]: we need to get the old value here (10000)

File: tests/manytcp-too-few-tls-vg.sh [hunk 10]
```diff
@@ -32,7 +34,7 @@ template(name="dynfile" type="string" string=`echo $RSYSLOG_OUT_LOG`) # trick to
 '
 startup_vg
 # the config file specifies exactly 1100 connections
-tcpflood -c1000 -m40000
+tcpflood -c1000 -m10000 -Ttls -x$srcdir/testsuites/x.509/ca.pem -Z$srcdir/testsuites/x.509/client-cert.pem -z$srcdir/testsuites/x.509/client-key.pem
```
Comment (rgerhards) [inline tests/manytcp-too-few-tls-vg.sh]: we need to get the old value here

File: tests/sndrcv_tls_certless_serveronly.sh [hunk 11]
```diff
@@ -1,19 +1,16 @@
 #!/bin/bash
-# all we want to test is if certless communication works. So we do
-# not need to send many messages.
-# This file is part of the rsyslog project, released under ASL 2.0
+# alorbach, 2019-01-16
+# This file is part of the rsyslog project, released  under ASL 2.0
 . ${srcdir:=.}/diag.sh init
-export NUMMESSAGES=5000
+export NUMMESSAGES=100
```
Comment (rgerhards) [inline tests/sndrcv_tls_certless_serveronly.sh:5]: we need to get the old value here

File: tests/sndrcv_tls_gtls_servercert_gtls_clientanon_legacy.sh [hunk 12]
```diff
@@ -4,7 +4,7 @@
 # This test checks legacy statements which are often given as 
 # This file is part of the rsyslog project, released under ASL 2.0
 . ${srcdir:=.}/diag.sh init
-export NUMMESSAGES=5000
+export NUMMESSAGES=1000
```
Comment (rgerhards) [inline tests/sndrcv_tls_gtls_servercert_gtls_clientanon_legacy.sh]: we need to get the old value here

File: tests/sndrcv_tls_gtls_servercert_gtls_clientanon_legacy.sh [hunk 13]
```diff
@@ -36,18 +35,24 @@ export RSYSLOG_DEBUGLOG="log2"
 generate_conf 2
 #export TCPFLOOD_PORT="$(get_free_port)"
 add_conf '
+global( defaultNetstreamDriverCAFile="'$srcdir/tls-certs/ca.pem'"
+)
+
 # Note: no TLS for the listener, this is for tcpflood!
-$ModLoad ../plugins/imtcp/.libs/imtcp
+module(	load="../plugins/imtcp/.libs/imtcp")
```
Comment (rgerhards) [inline tests/sndrcv_tls_gtls_servercert_gtls_clientanon_legacy.sh]: this needs to be done via legacy config statements, as before. The test checks if the legacy statements work ;-)  (thus the `-legacy` inside the test name).

File: tests/sndrcv_tls_gtls_servercert_gtls_clientanon_legacy.sh [hunk 14]
```diff
@@ -36,18 +35,24 @@ export RSYSLOG_DEBUGLOG="log2"
...
 
 # NOTE: do NOT change legacy statements - they are used intentionally
-$DefaultNetstreamDriverCAFile '$srcdir/tls-certs/ca.pem'
-$ActionSendStreamDriver gtls
-$ActionSendStreamDriverMode 1
-$ActionSendStreamDriverAuthMode x509/name
-$ActionSendStreamDriverPermittedPeer *.rsyslog.com
-*.* @@localhost:'$PORT_RCVR'
-
+# set up the action
```
Comment (rgerhards) [inline tests/sndrcv_tls_gtls_servercert_gtls_clientanon_legacy.sh]: this needs to be done via legacy config statements, as before. The test checks if the legacy statements work ;-)  (thus the `-legacy` inside the test name).

File: runtime/glbl.c [hunk 15]
```diff
@@ -96,7 +96,7 @@ static uchar * oversizeMsgErrorFile = NULL;		/* File where oversize messages are
 static int oversizeMsgInputMode = 0;	/* Mode which oversize messages will be forwarded */
 static int reportOversizeMsg = 1;	/* shall error messages be generated for oversize messages? */
 static int reportChildProcessExits = REPORT_CHILD_PROCESS_EXITS_ERRORS;
-static int iGnuTLSLoglevel = 0;
+static int iGnuTLSLoglevel = 2;		/* Sets GNUTLS Debug Level */
```
Comment (rgerhards) [inline runtime/glbl.c]: please restore the correct default value.

File: tests/imtcp-tls-ossl-basic.sh [hunk 16]
```diff
@@ -2,7 +2,7 @@
 # added 2018-04-27 by alorbach
 # This file is part of the rsyslog project, released under ASL 2.0
 . ${srcdir:=.}/diag.sh init
-export NUMMESSAGES=10000
+export NUMMESSAGES=50000
```
Comment (rgerhards) [inline tests/imtcp-tls-ossl-basic.sh]: previous value as 10000

**Comment by rgerhards** [comment]
doc PR: https://github.com/rsyslog/rsyslog-doc/pull/786

**Pull request merged**